### PR TITLE
fix: Stub generation testing and fixing of miscellaneous bugs

### PR DESCRIPTION
**Pull request by Masara**
### Summary of Changes

This PR contains the generated stub files from the [Library](https://github.com/Safe-DS/Library) project and is for testing those and fixing bugs, if they are found.

- [X] Collection[T] and Sequence[T] are recognized as TypeVars if used as superclasses
- [x] Mapping[] can be handled if used as superclass
- [x] Omit any class completely that inherits directly or transitively from ```Exception```
- [x] Rework how reexports are handled for the is_public method (currently not working correctly)
- [x] Fix stub package names: "The shortest public "reexport" should be used based on the ```__init__.py``` files."

### CODE REVIEW

**Comment by Masara** [comment]
@lars-reimann I added the generated stub files of the Library project to this PR branch and found the following:
- ```Collection[T]``` and ```Sequence[T]``` are not recognized as TypeVars, should I handle them the same way as the ```Generic``` class when used as a superclass?
- How should I handle ```Mapping``` as a superclass?
- In the file safeds.data.tabular.typing._column_type.sdsstub the keywords ```Boolean```, ```Nothing``` and ```String``` where marked as safeds keywords, so I added them to the keyword list which escapes them during the stub generation. Should I also add the other types? Also, they are still marked red, even after escaping ```Boolean```, ```Nothing``` and ```String``` in the sdsstub file.
- The stub for the class src/safeds/data/tabular/transformation/_discretizer shows an error for the class parameter ```numberOfBins: Float = 5```. The displayed error thinks that the value 5 is a literal, how should I handle this?
- Exception classes can't be recognized in safeds as superclasses, since the stubs don't know where they are imported from
- Furthermore, builtin classes like ```slice``` (see src/safeds/exceptions/_data.py:80 in the Library project) also can't be recognized by safeds, how should I display them? Or should I ignore them?

**Comment by lars-reimann** [comment]
> `Collection[T]` and `Sequence[T]` are not recognized as TypeVars, should I handle them the same way as the `Generic` class when used as a superclass?

Yes

> How should I handle `Mapping` as a superclass?

Remove any "builtin" superclasses like `Mapping`, `Collection`, or `Sequence`. Just the type vars are needed.

> Exception classes can't be recognized in safeds as superclasses, since the stubs don't know where they are imported from

You can omit any class completely that inherits directly or transitively from `Exception`.

> The stub for the class src/safeds/data/tabular/transformation/_discretizer shows an error for the class parameter `numberOfBins: Float = 5`. The displayed error thinks that the value 5 is a literal, how should I handle this?

An integer can't be assigned to a float. It should be `numberOfBins: Int = 5` here. That's just wrong in the library.

> In the file safeds.data.tabular.typing._column_type.sdsstub the keywords `Boolean`, `Nothing` and `String` where marked as safeds keywords, so I added them to the keyword list which escapes them during the stub generation. Should I also add the other types? Also, they are still marked red, even after escaping `Boolean`, `Nothing` and `String` in the sdsstub file.

These are not keywords, but core declarations of the language. Escaping won't help here, so please revert that. You can keep the names as they are. We'll need to manually give this another name and add a `@PythonName` annotation.

> Furthermore, builtin classes like `slice` (see src/safeds/exceptions/_data.py:80 in the Library project) also can't be recognized by safeds, how should I display them? Or should I ignore them?

Keep these types as they are for now.

**Comment by lars-reimann** [comment]
Also, the package names are wrong. The class `Column`, for example, should be in `safeds.data.tabular.containers`, not in `safeds.data.tabular.containers.Column`. The shortest public "reexport" should be used based on the `__init__.py` files.

**Comment by Masara** [comment]
@lars-reimann So if we have, for example
```py
def __init__(a: Mapping[str, Sequence[Any]): ...
```

should the type then be ```a: Map<str, Any>```?

**Comment by lars-reimann** [comment]
`Map<String, List<Any>>`

**Comment by Masara** [comment]
@lars-reimann So this code
```py
T = TypeVar('T')

class MappingTypeVar(Mapping[str, Any]):
    def __init__(self, data: Mapping[str, Any] | None = None): ...


class MappingTypeVar2:
    def __init__(self, data: Mapping[str, Sequence[Any]]): ...


class MappingTypeVar3(Mapping[str, Collection[T]]):
    def __init__(self, data: Mapping[str, Collection[T]] | None): ...
```
would translate to these stubs
```py
class MappingTypeVar(
	data: Map<String, Any?>?
)

class MappingTypeVar2(
	data: Map<String, List<Any>>
)

class MappingTypeVar3<List<T>>(
	data: Map<String, List<T>?>
)
```
?

**Comment by lars-reimann** [comment]
Almost:

```
class MappingTypeVar(
	data: Map<String, Any>? // Any, not Any?
)

class MappingTypeVar2(
	data: Map<String, List<Any>>
)

class MappingTypeVar3<T>( // T, not List<T>
	data: Map<String, List<T>>? // Moved the ? out. The map is nullable, not the value
)
```

**Comment by Masara** [comment]
@lars-reimann I am done with this PR, I updated the Library stub files in this branch so you can check them one last time to make sure everything is ok. If there are no problems, you can delete the Library-Stub directory in this branch, check the changes of this PR and merge.

**Comment by Masara** [comment]
@lars-reimann Nevermind, your [request regarding the package names](https://github.com/Safe-DS/Stub-Generator/pull/76#issuecomment-1977336985) still does not work according to your requirements. I thought that, for example, a module ```path/to/module``` which is exported in the ```path/to/__init__.py``` file like ```from path.to import module``` would get the stub text ```package path.to``` (only in this such a case). But according to you the package information is dependent on the class that is reexported in the ```__init__.py``` file, right? So if we have ```from path.to.module import SomeClass``` in the ```__init__.py``` instead of the first import I named, would we still get the stub text ```package path.to```? And what if we have other public classes in the module that are not imported in the ```__init__``` file?

**Comment by Masara** [comment]
But I'd say we close this PR first anyway, since the file changes are, once again, too much for a "quick" review.

**Comment by Masara** [comment]
So, after todays meeting there seemed to be a problem with these stubs. How would the stubs for the ```infer_alias_2 = AliasA``` part look like?

```py
from data.modules import ClassA as AliasA

class ClassB:
   ...

alias_b = ClassB

class AliasingModuleClassC():
   typed_alias: alias_b
   infer_alias = alias_b()
   typed_alias_2: AliasA
   infer_alias_2 = AliasA
```

```
@PythonModule("data.modules.aliasing_module_1")
package data.modules.aliasingModule1

from data.modules import ClassA

class ClassB()

class AliasingModuleClassC() {
    @PythonName("typed_alias")
    static attr typedAlias: ClassB
    @PythonName("infer_alias")
    static attr inferAlias: ClassB
    @PythonName("typed_alias_2")
    static attr typedAlias2: ClassA
    @PythonName("infer_alias_2")
    static attr inferAlias2: ClassA
}
```

**Comment by lars-reimann** [comment]
> @lars-reimann Nevermind, your [request regarding the package names](https://github.com/Safe-DS/Stub-Generator/pull/76#issuecomment-1977336985) still does not work according to your requirements. I thought that, for example, a module `path/to/module` which is exported in the `path/to/__init__.py` file like `from path.to import module` would get the stub text `package path.to` (only in this such a case). But according to you the package information is dependent on the class that is reexported in the `__init__.py` file, right? So if we have `from path.to.module import SomeClass` in the `__init__.py` instead of the first import I named, would we still get the stub text `package path.to`? And what if we have other public classes in the module that are not imported in the `__init__` file?

Ask yourself how you can import a class/global function from a library into your **Python** code. There might be multiple ways to do so, if they get reexported in `__init__.py` files. Out of the possible ways, pick the _public_ module path that is

1. a subpath of the path to the module that declares the class/global function,
2. the shortest possible one.

**Comment by lars-reimann** [comment]
> So, after todays meeting there seemed to be a problem with these stubs. How would the stubs for the `infer_alias_2 = AliasA` part look like?
> 
> ```python
> from data.modules import ClassA as AliasA
> class AliasingModuleClassC():
>    infer_alias_2 = AliasA
> ```

We cannot represent the type of `infer_alias_2` in the stubs, since we don't have a use case for reflection. Just omit the type and put a TODO above that line.

File: tests/safeds_stubgen/stubs_generator/__snapshots__/test_generate_stubs/TestStubFileGeneration.test_stub_creation[_module_3].sdsstub [hunk 1]
```diff
@@ -1,4 +1,4 @@
-@PythonModule("tests.data.various_modules_package.file_creation._module_3")
-package tests.data.variousModulesPackage.fileCreation.Module3
+@PythonModule("various_modules_package")
+package variousModulesPackage
```
Comment (lars-reimann) [inline tests/safeds_stubgen/stubs_generator/__snapshots__/test_generate_stubs/TestStubFileGeneration.test_stub_creation[_module_3].sdsstub:2]: The removal of `file_creation._module_3` is correct, but the prefix `tests.data` should remain in this case.

Comment (lars-reimann) [reply tests/safeds_stubgen/stubs_generator/__snapshots__/test_generate_stubs/TestStubFileGeneration.test_stub_creation[_module_3].sdsstub:2]: The same applies to the other tests here.

**Review by lars-reimann — APPROVED** [review]
Great work, thank you.

**Pull request merged**